### PR TITLE
Positional encoding

### PR DESCRIPTION
**Pull request by kumarutkarsh1248**
This PR adapts the PositionalEncoding layer.
It is a non-trainable layer that adds sinusoidal positional encodings to the embedding output.
This layer has a private helper function, `InitPositionalEncoding()` which initializes the `positionalEncoding` values.

Current implementation calls `InitPositionalEncoding() `inside the `Forward()` once during first forward pass and afterward throughout the training and inference `positionalEncoding` values remain unchanged.

### CODE REVIEW

File: src/mlpack/methods/ann/layer/positional_encoding.hpp [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+ * position of the tokens in the sequence.
+ *
+ * The input and the output have the same shape:
+ * `(embedDim * maxSequenceLength, batchSize)`. The embeddings are stored
+ * consequently.
+ *
+ * @tparam InputType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding.hpp:33]: Are these comments correct?  I think that the input and output types have to be matrix types, but they can hold any element type.

File: src/mlpack/methods/ann/layer/positional_encoding.hpp [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+  /**
+   * Create the PositionalEncoding layer object using the specified parameters.
+   *
+   * @param embedDim The length of the embedding vector.
+   * @param maxSequenceLength Number of tokens in each sequence.
+   */
+  PositionalEncoding(const size_t embedDim,
+                    const size_t maxSequenceLength,
+                   RegularizerType regularizer = RegularizerType());
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding.hpp:56]:
```suggestion
                     const size_t maxSequenceLength,
                     RegularizerType regularizer = RegularizerType());
```

File: src/mlpack/tests/ann/layer/positional_encoding.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+//   input = arma::linspace<arma::rowvec>(0, 9, 10);
+//   output.set_size(embeddingSize, 10);
+//   e.Forward(input, output);
+
+//   REQUIRE(output.n_rows == embeddingSize);
+//   REQUIRE(output.n_cols == input.n_cols);
+//   for (size_t i = 0; i < 10; ++i)
+//     REQUIRE(approx_equal(output.col(i), weights.col(i), "absdiff", 1e-5));
+// }
```
Comment (rcurtin) [inline src/mlpack/tests/ann/layer/positional_encoding.cpp:85]: I think this is adapted from the embedding layer... should we remove it?

File: src/mlpack/methods/ann/layer/positional_encoding_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    const MatType& error,
+    MatType& gradient)
+{
+  // Nothing to do here.
+}
+
+template<typename MatType, typename RegularizerType>
+void PositionalEncoding<MatType, RegularizerType>::ComputeOutputDimensions()
+{
+  this->outputDimensions = this->inputDimensions;
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding_impl.hpp:153]: How should this layer apply to 2-dimensional inputs?  If e.g. the user passes in an image, inputDimensions might be `{ width, height }`, but we are applying positional encoding linearly to each element.  Is that the right thing to do when the input is in higher dimensions?  If you want to restrict this layer to cases where `inputDimensions` is only 1-dimensional (which I think is the general use case for positional encoding) I think that's fine too.

File: src/mlpack/methods/ann/layer/positional_encoding_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    positionalEncoding.col(2 * i + 1) = arma::cos(theta.col(i));
+  }
+  positionalEncoding = vectorise(positionalEncoding.t());
+}
+
+template<typename MatType, typename RegularizerType>
+void PositionalEncoding<MatType, RegularizerType>::Forward(
+    const MatType& input, MatType& output)
+{
+  if (positionalEncoding.n_elem == 0) InitPositionalEncoding();
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding_impl.hpp:123]: I would instead check that `positionalEncoding.n_rows != input.n_rows`, because it is possible that a `positionalEncoding` could have been set earlier, then the network changed size, then `Forward()` gets called again.  That is of course an unusual situation, but we should still be aware that in any `Forward()` call, the input size is not guaranteed to be the same as it was last time.  Any network size changes will be done with a call to `SetOutputDimensions()`, so I actually wonder if calling `InitPositionalEncoding()` would be better done there.

File: src/mlpack/methods/ann/layer/positional_encoding.hpp [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+  //! Serialize the layer.
+  template<typename Archive>
+  void serialize(Archive& ar, const uint32_t /* version */);
+
+ private:
+  /**
+   * Initialize positional encodings for further use.
+   */
+  void InitPositionalEncoding();
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding.hpp:130]: So there is a tradeoff with this strategy: by using `InitPositionalEncoding()`, we precompute the positional encodings so that `Forward()` is really easy to implement, but the cost is that we use extra memory at inference time.

But, Armadillo is all about delayed evaluation and inlining operations.  So for instance, if I do

```c++
arma::mat B = arma::regspace<arma::mat>(...);
output = input + B;
```

and also

```
output = input + arma::regspace<arma::mat>(...);
```

the second will actually be more efficient because Armadillo is able to inline the generation of the `regspace` elements directly into the loop over the elements of `output`, without ever forming an auxiliary matrix `B` (so e.g. the cost of allocating memory, doing the copy, etc., is not there; in addition, when we set `output`, there is less memory pressure since we are generating the `regspace` elements on the fly).  Armadillo can actually do this with fairly complex operations.  So, I am wondering the code might actually be faster if you inlined the computation of `positionalEncoding` directly into the `Forward()` pass, without using any temporaries (e.g. theta, position, divTerm, etc.).  That implementation would no longer be a simple `+` operation, but I do think it would be faster overall.

Are you willing to give that a shot?  You could benchmark it in a simple test program to see if my intuition is off or not.

File: src/mlpack/methods/ann/layer/positional_encoding_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+template<typename MatType, typename RegularizerType>
+void PositionalEncoding<MatType, RegularizerType>::Gradient(
+    const MatType& input,
+    const MatType& error,
+    MatType& gradient)
+{
+  // Nothing to do here.
+}
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding_impl.hpp:148]: You can actually just omit this function and the empty base function from `Layer()` will be used instead.

**Review by rcurtin — COMMENTED** [review]
Thanks for working on this!  I noticed a few things looking through; let me know what you think.

**Comment by kumarutkarsh1248** [comment]
I did some research, and now I think we don’t need a separate positional encoding layer. We can make the embedding layer behave like a positional encoding layer.

The idea is that positional encoding comes in different forms: trainable and non-trainable. Even within non-trainable types, there are multiple variants such as sinusoidal and Fourier. So instead of creating a dedicated layer for each of these forms, we can reuse the embedding layer to serve as positional encoding, since positional encoding is essentially just a lookup table.

In fact, major libraries like PyTorch and TensorFlow follow a similar approach.

Let me give a demonstration here:
Screenshot from 2026-04-05 17-34-25

we can make this using DagNetwork:

```cpp
#include <mlpack.hpp>
#include <armadillo>
#include <iostream>

using namespace mlpack;
using namespace std;

// A layer that turns token indices into position indices.
// Example:
// input  shape: (seqLen, batchSize)
// output shape: (seqLen, batchSize)
//
// If seqLen = 4 and batchSize = 2, output becomes:
// [0 0
//  1 1
//  2 2
//  3 3]
class PositionLayer : public mlpack::Identity<arma::mat>
{
 public:
  PositionLayer() {}

  void Forward(const arma::mat& input, arma::mat& output)
  {
    const size_t rows = input.n_rows;
    const size_t cols = input.n_cols;

    arma::vec pos = arma::regspace<arma::vec>(0, rows - 1);
    output = arma::repmat(pos, 1, cols);
  }
};

int main()
{
  const size_t vocabSize = 20;
  const size_t embeddingSize = 8;
  const size_t seqLen = 4;
  const size_t batchSize = 2;

  mlpack::DAGNetwork<> model;
  model.InputDimensions() = { seqLen };

  const size_t l1 = model.Add<mlpack::Identity>();
  const size_t l2 = model.Add<PositionLayer>();
  const size_t l3 = model.Add<mlpack::Embedding<>>(vocabSize, embeddingSize);
  const size_t l4 = model.Add<mlpack::Embedding<>>(seqLen, embeddingSize);
  const size_t l5 = model.Add<mlpack::Identity>();

  // Graph connections.
  model.Connect(l1, l3);  // input tokens -> token embedding
  model.Connect(l1, l2);  // input tokens -> position generator
  model.Connect(l2, l4);  // positions -> positional embedding
  model.Connect(l3, l5);  // token embedding -> add
  model.Connect(l4, l5);  // positional embedding -> add
  model.SetConnection(l5, ADDITION);
  model.Reset();

  arma::mat input(seqLen, batchSize);
  input.col(0) = arma::vec({1, 5, 2, 7});
  input.col(1) = arma::vec({3, 4, 8, 6});

  arma::mat output;
  model.Forward(input, output);

  cout << "Input token ids:\n" << input << "\n";
  cout << "Output shape: " << output.n_rows << " x " << output.n_cols << "\n";
  cout << "Output:\n" << output << "\n";

  return 0;
}
```

This way positional encoding layer will behave as an trainable layer and if we want to use some custom positional encoding values then we can set the weight of embedding layer accordingly.
@rcurtin @andrewfurey21 please let me know if you think this can be done in a better way.

so for now, I think we just need to remove `PositionalEncoding` Layer from the `not_adapted` folder as it doesn’t seem necessary anymore.

**Comment by rcurtin** [comment]
I am trying to understand the overall goal here because I believe what you wrote above does not match the implementation of the `PositionalEncoding` layer as I read it.  I think what you are saying is one of the following two things; can you tell me which is right? (Or if none of the above... then let me know what it is :smile: )

1. The `PositionalEncoding` layer is intended to be used in conjunction with an `Embedding` layer; the graph shows the desired use as part of a larger neural network; or,

2. The `PositionalEncoding` layer's operation is exactly represented as the depicted graph.

I can give a better recommendation if I understand the overall goal.  I will say though, that representing the layer as an entire `DAGNetwork` will end up being really inefficient.  It would be way better to simply implement the entire set of operations you showed in the figure in a single layer's `Forward()` or `Backward()` pass.

**Comment by kumarutkarsh1248** [comment]
Sorry for the confusion. I think the fact that I first opened the PR for a `PositionalEncoding` layer and then later said we may not need it is caused the confusion 😄


> The PositionalEncoding layer's operation is exactly represented as the depicted graph.

Yes, The graph above shows how the combined token embedding + positional information can be represented using `DAGNetwork`.

So from a functionality point of view, we do not strictly need a separate `PositionalEncoding` layer. We can use an `Embedding` layer together with the DAG structure above to reproduce the same behavior.

For reference:
In PyTorch and TensorFlow, it is a standard practice to use an embedding layer in place of positional encoding and construct a graph like the one above. Therefore, these two major libraries do not have a dedicated positional encoding layer.

But we can still keep a `PositionalEncoding` layer in mlpack since the implementation is almost done. My only concern is that this layer would represent one very specific variant only: fixed, non-trainable sinusoidal positional encoding. There are many other forms of positional encoding, and I dont think we would want to keep adding all such variants directly into the codebase as separate layers.

For my use case, I can go ahead with `DAGNetwork` approach.

**Comment by rcurtin** [comment]
Thanks for the clarification.  I don't have a problem keeping the `PositionalEncoding` layer in mlpack, if there is a reasonable chance that a user would actually use it when building a network.  What do you think?

If you do end up implementing the network in the picture with DAGNetwork and have efficiency issues, I think it would be reasonable to add it all as a specially-implemented single layer to mlpack.

File: src/mlpack/methods/ann/layer/positional_encoding.hpp (hunk 6 above)
Comment (kumarutkarsh1248) [reply src/mlpack/methods/ann/layer/positional_encoding.hpp:130]: > output = input + arma::regspace<arma::mat>(...);

I can implement this, but the problem is that in my case I will be using trained positional encoding values, which are not sinusoidal. So I need some way to do something like `mlpack::PositionalEncoding<>.Parameters() = customWeights;`

That is why I thought that if the layer stores the internal encoding values, then we can modify them accordingly.

**Comment by kumarutkarsh1248** [comment]
> I don't have a problem keeping the PositionalEncoding layer in mlpack

Okay, then I will complete this PR 

> positional encoding comes in different forms: trainable and non-trainable. Even within non-trainable types, there are multiple variants such as sinusoidal and Fourier.

I have an idea: by using a trainable positional encoding layer, we can obtain the behavior of all the different variants of positional encoding. However, this requires one feature from mlpack::FFN: the ability to freeze the weights of some internal layers.

For example, if we have layer1 -> layer2 -> layer3, I need some way to freeze the weights of layer2 so that during training, the weights of layer1 and layer3 change, but the weights of layer2 remain the same.

If we can do this, then to make the trainable positional encoding layer behave like a sinusoidal positional encoding layer, the user can simply set the weights of the trainable positional layer to sinusoidal weights and freeze that layer’s weights.

Now the question is: can we freeze the weights of some layer during training? I think later, when we fine-tune BERT and YOLO in mlpack, we are going to require this property.

**Comment by andrewfurey21** [comment]
> Now the question is: can we freeze the weights of some layer during training? I think later, when we fine-tune BERT and YOLO in mlpack, we are going to require this property.

Probably not that hard. We could add a  `std::vector<bool>` to FFN and DAGNetwork maybe? And just skip computing gradients on those layers.

**Comment by rcurtin** [comment]
> Probably not that hard. We could add a std::vector<bool> to FFN and DAGNetwork maybe? And just skip computing gradients on those layers.

I agree with that design.  At the level of ensmallen, the frozen parameters will still be a part of the matrix that `Evaluate()` keeps track of, but keeping the gradient of all those elements as zero will mean that the elements should never be changed (at least for most optimizers).

The other design option would be to have FFN/DAGNetwork/RNN hold two different parameter matrices: one for frozen parameters, one for regular trainable parameters.  Then `SetLayerMemory()` would have to just be a little bit careful about which one it used.  However, that would guarantee that *no* ensmallen optimizer would ever modify non-trainable weights, since they aren't even accessible to the optimizer.  It would also be very incrementally faster because the coordinates matrix ensmallen was using would be just a bit smaller.  (Likely usually a negligible speedup.)

Hopefully what I am writing makes sense---I know I am writing a little vaguely.  If you want me to clarify what I mean just let me know.

**Comment by kumarutkarsh1248** [comment]
> The other design option would be to have FFN/DAGNetwork/RNN hold two different parameter matrices: one for frozen parameters, one for regular trainable parameters.

I think this approach is better bcz, during the fine-tuning of BERT, a large number of weights are goind to be frozen.

To give an idea, a fine-tuned BERT model usually looks like: `base BERT model + an additional layer.` BERT itself has around 100 million parameters, which I would keep frozen during fine-tuning, and only the weights of the additional layers would be trained.

So, I believe fine-tuning would be much faster if we go with the second approach.

**Comment by rcurtin** [comment]
@kumarutkarsh1248 just wanted to check in on this one.  Did you want to finish this positional encoding layer?  I'm ready to review/etc., just let me know.

**Comment by kumarutkarsh1248** [comment]
@rcurtin, could you please take a look at the `Forward()` pass implementation? Do you think we can optimize it further?

Also, I think I’m definitely missing something here, because with this approach we would be calculating the positional encoding during every forward pass. Previously, I was keeping a global `positionalEncoding` matrix that was computed only once during the first forward pass.

File: src/mlpack/methods/ann/layer/positional_encoding_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        (-std::log(10000.0) / embedDim)
+    ).t();
+
+  arma::uvec evenCols = arma::regspace<arma::uvec>(0, 2, embedDim - 2);
+  arma::uvec oddCols  = arma::regspace<arma::uvec>(1, 2, embedDim - 1);
+
+  positionalEncoding.cols(evenCols) = arma::sin(theta);
+  positionalEncoding.cols(oddCols)  = arma::cos(theta);
+
+  output = input.each_col() + arma::vectorise(positionalEncoding.t());
```
Comment (rcurtin) [inline src/mlpack/methods/ann/layer/positional_encoding_impl.hpp:118]: So my belief is that we don't need the `positionalEncoding` matrix at all, and we can just perform the operation in-place.  You could try a proof-of-concept by not creating the `theta` matrix, and writing a loop like this:

```
for (size_t c = 0; c < input.n_cols; ++c)
{
  for (size_t r = 0; r < input.n_rows; ++r)
  {
    output(r, c) = input(r, c) + (a scalar expression that is the right element in positional encoding)
  }
}
```

Since I expect the code to be memory-bound instead of compute-bound, this is likely to be a good bit faster than precomputing the positional encoding matrix (and it reduces the amount of memory necessary to run the network, which is also important).

If you can make that work, then we can put a little bit of effort into figuring out how to vectorize it nicely with Armadillo expressions.  And if we can't vectorize it, then there are some other layers that use OpenMP loops to set the elements individually, and we can just do that here.

**Review by rcurtin — COMMENTED** [review]
@kumarutkarsh1248 I am really sorry it took me so long to get back to this.  It has not been a friendly month for me.  I left a comment that I hope is clarifying.  You have spent more time in this code than I have, so it's always possible that I've totally overlooked something.  I should be able to stay more on top of any responses to this PR now (sorry again!).
